#! /bin/sh -e

# All lines beginning with `# DP:' are a description of the patch.
# DP: Description: Support libraries that has execstack on <= 2.4.18 kernels.
# DP: Related bugs: #321718
# DP: Dpatch author: GOTO Masanori <gotom@debian.org>
# DP: Patch author: GOTO Masanori <gotom@debian.org>
# DP: Upstream status: Not submitted
# DP: Status Details: Will be submitted
# DP: Date: 2005-08-19

PATCHLEVEL=0

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p$PATCHLEVEL < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p$PATCHLEVEL < $0;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1
esac
exit 0

# append the patch here and adjust the -p? flag in the patch calls.
2005-08-19  GOTO Masanori  <gotom@debian.org>

	* sysdeps/unix/sysv/linux/dl-execstack.c: Check not only ENOMEM but
	also EFAULT for pre 2.4.19 kernels.

--- sysdeps/unix/sysv/linux/dl-execstack.c.old	2006-01-28 22:25:42.000000000 -0500
+++ sysdeps/unix/sysv/linux/dl-execstack.c	2006-01-28 22:26:06.000000000 -0500
@@ -89,7 +89,7 @@
 	page -= size;
       else
 	{
-	  if (errno != ENOMEM)	/* Unexpected failure mode.  */
+	  if (errno != ENOMEM && errno != EFAULT)	/* Unexpected failure mode.  */
 	    {
 	      result = errno;
 	      goto out;
@@ -115,7 +115,7 @@
 	page += size;
       else
 	{
-	  if (errno != ENOMEM)	/* Unexpected failure mode.  */
+	  if (errno != ENOMEM && errno != EFAULT)	/* Unexpected failure mode.  */
 	    {
 	      result = errno;
 	      goto out;
